OpenRA.Mods.Common/DiscordService.cs

```diff
@@ -0,0 +1,200 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using DiscordRPC;
+using DiscordRPC.Message;
+using OpenRA.Network;
+
+namespace OpenRA.Mods.Common
+{
+	public enum DiscordState
+	{
+		Unknown,
+		InMenu,
+		InMapEditor,
+		InSkirmishLobby,
+		InMultiplayerLobby,
+		PlayingMultiplayer,
+		WatchingReplay,
+		PlayingCampaign,
+		PlayingSkirmish
+	}
+
+	public class DiscordService : IGlobalModData, IDisposable
+	{
+		public readonly string ApplicationId = null;
+		readonly DiscordRpcClient client;
+		DiscordState currentState;
+
+		static readonly Lazy<DiscordService> Service = Exts.Lazy(() =>
+		{
+			if (!Game.ModData.Manifest.Contains<DiscordService>())
+				return null;
+
+			return Game.ModData.Manifest.Get<DiscordService>();
+		});
+
+		public DiscordService(MiniYaml yaml)
+		{
+			FieldLoader.Load(this, yaml);
+
+			client = new DiscordRpcClient(ApplicationId, autoEvents: false)
```

Comment:
From IRC: It sounds like this is also running on dedicated servers (and probably the utility too), which would be bad. We should check and abort here if it is not a normal game session - e.g. if `Game.Renderer == null`.